slack-code-runner-bot/function.js

```diff
@@ -0,0 +1,123 @@
+const { WebClient } = require('@slack/client');
+const { VM } = require('vm2');
+const decode = require('decode-html');
+
+const {
+  SLACK_BOT_TOKEN: token,
+  BOT_TRIGGER_STATEMENT: triggerStatement,
+  MAX_CODE_RUNTIME_MS: maxTimeMS,
+} = process.env;
+const web = new WebClient(token);
+
+const triBacktick = '```';
+
+/**
+ * Extracts the potential JavaScript code segment located
+ * between leading and trailing triple backticks ```
+ *
+ * @param {string} textToExtractFrom - segment to attempt code extraction on
+ *
+ * @return {string} - extracted code segment
+ */
+function extractBetweenBackticks(textToExtractFrom) {
+  const startIdx = textToExtractFrom.indexOf(triBacktick);
+  const endIdx = textToExtractFrom.lastIndexOf(triBacktick);
+
+  if (startIdx === -1 || endIdx === -1 || (startIdx === endIdx)) {
+    throw new Error('Javascript input must begin and end with triple backticks');
+  }
+  return extractText.substring(startIdx + 3, endIdx);
+}
+
+/**
+ * Runs the provided code and returns computed output.
+ *
+ * @param {string} code - Javascript code to evaluate
+ * @param {Array} logBuffer - buffer to insert stdout ouput into
+ *
+ * @return {string} - output of evaluated code
+ */
+async function runCode(code, logBuffer) {
+  // replacement for logic usually handled by "console"
+  const pushLogs = (data) => logBuffer.push(data);
+  const redirectConsole = {
+    err: pushLogs,
+    log: pushLogs,
+    warn: pushLogs,
+  };
+
+  let result;
+  const vm = new VM({
+    timeout: maxTimeMS,
+    sandbox: { console: redirectConsole },
+    require: { external: true },
```

Comment:
How safe is this?  *Any* external module?  (I can install what I like if I use `child_process` to run `npm`, or if I just download the code with `http` and require it).

Reply:
I honestly am not 100% sure based on the docs. We can air on the safe side and initially just block `require`.